OpenRA.Mods.Common/Activities/Air/FlyAttack.cs

```diff
@@ -150,14 +150,39 @@ public override Activity Tick(Actor self)
 				return this;
 			}
 
-			if (self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length < aircraft.Info.MinAirborneAltitude)
-				QueueChild(self, new TakeOff(self), true);
+			var delta = attackAircraft.GetTargetPosition(pos, checkTarget) - pos;
+			var desiredFacing = delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : aircraft.Facing;
+			var isAirborne = self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length >= aircraft.Info.MinAirborneAltitude;
+
+			if (!aircraft.Info.CanHover)
+			{
+				if (!isAirborne)
+					QueueChild(self, new TakeOff(self), true);
 
-			if (attackAircraft != null && target.IsInRange(self.CenterPosition, attackAircraft.GetMinimumRange()))
-				QueueChild(self, new FlyTimed(ticksUntilTurn, self), true);
+				if (attackAircraft != null && target.IsInRange(self.CenterPosition, attackAircraft.GetMinimumRange()))
+					QueueChild(self, new FlyTimed(ticksUntilTurn, self), true);
 
-			QueueChild(self, new Fly(self, target, checkTarget.CenterPosition, Color.Red), true);
-			QueueChild(self, new FlyTimed(ticksUntilTurn, self));
+				QueueChild(self, new Fly(self, target, checkTarget.CenterPosition, Color.Red), true);
+				QueueChild(self, new FlyTimed(ticksUntilTurn, self));
+			}
+			else if (!isAirborne)
+				QueueChild(self, new TakeOff(self), true);
+			else if (!target.IsInRange(pos, lastVisibleMaximumRange))
+				QueueChild(self, new Fly(self, target, WDist.Zero, lastVisibleMaximumRange, checkTarget.CenterPosition, Color.Red), true);
+			else if (target.IsInRange(pos, attackAircraft.GetMinimumRangeVersusTarget(target)))
+			{
+				// Facing 0 doesn't work with the following position change
```

Comment:
What is this code trying to do? Remember that facing 0 and 1 are real facings, -1 is used to mean "any facing".